python/cudf/cudf/pylibcudf_tests/test_hashing.py

```diff
@@ -0,0 +1,222 @@
+                    "Some multi-byte characters here: ééé",
+                    "ééé",
+                    "ééé ééé",
+                    "ééé ééé ééé ééé",
+                    "",
+                    "!@#$%^&*(())",
+                    "0123456789",
+                    "{}|:<>?,./;[]=-",
+                ],
+                type=pa.string(),
+            )
+        ],
+        names=["data"],
+    )
+    return plc.interop.from_arrow(arrow_tbl)
+
+
+@pytest.fixture(scope="module")
+def xxhash_64_decimal_tbl():
+    arrow_tbl = pa.Table.from_arrays(
+        [pa.array([0, 100, -100, 999999999, -999999999], type=pa.decimal(-3))],
+        names=["data"],
+    )
+    return plc.interop.from_arrow(arrow_tbl)
+
+
+# Full table hash
+@pytest.fixture(scope="module")
+def all_types_input_table():
+    data = pa.Table.from_pydict(
+        {
+            "int": [1, 2, 3],
+            "float": [1.0, 2.0, 3.0],
+            "bool": [True, False, True],
+            "string": ["a", "b", "c"],
+            "list": [[1], [2], [3]],
+            "struct": [{"a": 1}, {"a": 2}, {"a": 3}],
+        }
+    )
+    return data
+
+
+def all_types_output_table(input, method):
+    def _applyfunc(x):
+        hasher = getattr(hashlib, method)
+        # TODO: not how libcudf computes row hash
+        return hasher(str(x).encode()).hexdigest()
+
+    result = pa.Table.from_pandas(input.to_pandas().map(_applyfunc))
+    return result
+
+
+@pytest.mark.parametrize("method", METHODS)
+def test_hash_column(pa_input_column, method):
+    def _applyfunc(x):
+        hasher = getattr(hashlib, method)
+        return hasher(str(x).encode()).hexdigest()
+
+    plc_tbl = plc.interop.from_arrow(
+        pa.Table.from_arrays([pa_input_column], names=["data"])
+    )
+    plc_hasher = getattr(plc.hashing, method)
+
+    expect = pa.Array.from_pandas(
+        pa_input_column.to_pandas().apply(_applyfunc)
+    )
+    got = plc_hasher(plc_tbl)
+    assert_column_eq(got, expect)
+
+
+@pytest.mark.parametrize(
+    "method", ["sha1", "sha224", "sha256", "sha384", "sha512"]
+)
+@pytest.mark.parametrize("dtype", ["list", "struct"])
+def test_sha_list_struct_err(all_types_input_table, dtype, method):
+    err_types = all_types_input_table.select([dtype])
+    plc_tbl = plc.interop.from_arrow(err_types)
+    plc_hasher = getattr(plc.hashing, method)
+
+    with pytest.raises(ValueError):
+        plc_hasher(plc_tbl)
+
+
+def test_xxhash_64_int(xxhash_64_int_tbl):
+    expected = pa.array(
+        [
+            4827426872506142937,
+            13867166853951622683,
+            4246796580750024372,
+            17339819992360460003,
+            7292178400482025765,
+            2971168436322821236,
+            9380524276503839603,
+            9380524276503839603,
+        ],
+        type=pa.uint64(),
+    )
+    got = plc.hashing.xxhash_64(xxhash_64_int_tbl, 0)
+    assert_column_eq(got, expected)
```

Comment:
We discussed a bit in calls, but do tests like this one give us any additional coverage beyond what the libcudf tests give us? I'm wondering if the more generic tests using e.g. the `all_types_input_table` would be sufficient for coverage on the Python side.

Reply:
For these I tried to copy the tests from [the libcudf test suite](https://github.com/rapidsai/cudf/blob/branch-24.06/cpp/tests/hashing/xxhash_64_test.cpp#L58) as faithfully as possible, so at least for this algorithm we should have 1-1 coverage.

Reply:
Right, what I mean is are we doing redundant work by testing exact things that libcudf is already testing for us?

Reply:
I think it's redundant in the sense that we don't test anything _additional_, but I think testing the same exact thing except from python serves as a fairly robust indicator that our bindings aren't doing anything wrong (we can be at least as sure as `libcudf` is of itself)

Reply:
My point is that pylibcudf testing should essentially just be about verifying that the bindings themselves aren't doing anything wrong. The rough decision framework I would use is:
1. If this test fails, could it indicate a bug in libcudf? If so, the test should be rewritten as a libcudf gtest if one does not already exist.
2. If the test could indicate a bug in pylibcudf, can we rewrite the test in such a way as to isolate the pylibcudf behavior and avoid capturing libcudf issues?
 
IOW, we want to maximize the chances that if libcudf's test suite is passing, any failures in pylibcudf's test suite are strictly indicative of issues with the bindings (notwithstanding special cases like I/O where we might choose to test only in Python rather than C++ because it's hard to construct test cases). We want to accomplish this while minimizing the size of the pylibcudf test suite.

Right now this test seems like it could indicate a bug in libcudf or pylibcudf. However, we have a corresponding libcudf test, so if libcudf's test suite is passing and this test fails, that would indicate an issue in the bindings. The question is, would such binding issues be more easily identified using a simpler test? Hashing is a somewhat odd case for cudf in that it's essentially a mathematical operation with clear inputs and outputs, so it's fairly easy to generate gold data. If in your estimation the easiest way to write a minimal test of pylibcudf bindings is to just copy gold data from the libcudf test and run it here, that's totally fine with me. I just want to make sure that we're framing the problem here as "what is the minimal unit test that captures pylibcudf-specific errors".

CC @bdice

Reply:
```
so if libcudf's test suite is passing and this test fails, that would indicate an issue in the bindings.
```
I think this articulates the intuition I had behind the idea. This divergence feels like the thing we're trying to capture with these tests, so if there's a simpler way of achieving that I am for it. However I agree it maybe is a little weird that the information comes from considering both tests at once.

Reply:
As I figured out more ways of hashing things in python, I ended up removing this. The updated code should dodge around the need for this explicitly going forward.